File: engine.py
from evaluation import (
    evaluate_move,
    is_endgame,
    evaluate,
)

from chess import (
    # Squares, Boards, and Moves.
    Board,
    Move,

    # Color.
    WHITE
)

def get_ordered_moves(captures: bool, board: Board) -> list[Move]:
    """
        Retrieves the ordered moves with the `evaluate_move` function.
    """

    endgame = is_endgame(board)

    ordered_moves = sorted(
        board.legal_moves,
        reverse=(board.turn == WHITE),
        key=lambda move: evaluate_move(move, endgame, board),
    )

    if captures:
        return [move for move in ordered_moves if board.is_capture(move)]

    return ordered_moves

def quiesce(alpha: float, beta: float, board: Board) -> float:
    """
        This function loops over all the captures until no captures are left, then returns the evaluation of the position with no captures.
    """

    evaluation = evaluate(board)
    if evaluation >= beta:
        return beta

    alpha = max(evaluation, alpha)
    captures = get_ordered_moves(True, board)

    for capture in captures:
        board.push(capture)
        evaluation = -quiesce(-beta, -alpha, board)

        board.pop()
        if evaluation >= beta:
            return beta
        
        alpha = max(evaluation, alpha)
    
    return alpha

def negamax(
    depth: int,
    alpha: float,
    beta: float,
    board: Board,
) -> float:
    """
        Evaluates a position to a specific depth with Minimax, Alpha-Beta Pruning and Move Ordering.
    """

    depth = max(depth, 0)
    if board.is_checkmate():
        return float('-inf')
    elif board.is_game_over():
        return 0.0
    
    if depth == 0:
        return quiesce(alpha, beta, board)

    moves = get_ordered_moves(False, board)

    for move in moves:
        board.push(move)
        evaluation = -negamax(
            depth - 1,
            -beta,
            -alpha,
            board
        )

        board.pop()
        if evaluation >= beta:
            return beta
        
        alpha = max(evaluation, alpha)

    return alpha

def find_best_move(
    depth: int,
    board: Board,
) -> tuple[Move, float]:
    """
        Finds the best move with the `negamax` function.
    """

    moves = get_ordered_moves(False, board)
    best_evaluation = float('-inf')

    for move in moves:
        board.push(move)
        if board.can_claim_draw():
            evaluation = 0.0
        else:
            evaluation = -negamax(
                depth - 1,
                float('-inf'),
                float('inf'),
                board,
            )

        board.pop()
        if best_evaluation <= evaluation:
            best_evaluation = evaluation
            best_move = move

    return best_move, best_evaluation

File: engine.go
package main

import (
	"fmt"
	"math"
	"sort"

	"github.com/notnil/chess"
)

func orderMoves(moves []*chess.Move, position *chess.Position) {
	var endgame = isEndgame(position)

	sort.Slice(moves, func(i, j int) bool {
		return evaluateMove(moves[i], endgame, position) < evaluateMove(moves[j], endgame, position)
	})

	if position.Turn() == chess.White {
		sort.SliceStable(moves, func(_, _ int) bool {
			return true
		})
	}
}

func quiesce(alpha float64, beta float64, position *chess.Position) float64 {
	var evaluation float64 = evaluate(position)

	if evaluation >= beta {
		return beta
	}

	alpha = math.Max(evaluation, alpha)
	var moves []*chess.Move = position.ValidMoves()
	var captures []*chess.Move

	for _, move := range moves {
		if move.HasTag(chess.Capture) {
			captures = append(captures, move)
		}
	}

	orderMoves(captures, position)

	for _, capture := range captures {
		var newPosition *chess.Position = position.Update(capture)

		evaluation = -quiesce(-beta, -alpha, newPosition)

		if evaluation >= beta {
			return beta
		}

		alpha = math.Max(evaluation, alpha)
	}

	return alpha
}

func negamax(depth uint8, alpha float64, beta float64, position *chess.Position) float64 {
	var moves []*chess.Move = position.ValidMoves()

	if position.Status() == chess.Checkmate {
		return math.Inf(-1)
	}

	if len(moves) == 0 {
		return 0.0
	}

	if depth == 0 {
		return quiesce(alpha, beta, position)
	}

	orderMoves(moves, position)

	for _, move := range moves {
		var newPosition *chess.Position = position.Update(move)
		var evaluation float64 = -negamax(depth-1, -beta, -alpha, newPosition)

		if evaluation >= beta {
			return beta
		}

		alpha = math.Max(evaluation, alpha)
	}

	return alpha
}

func findBestMove(depth uint8, game *chess.Game) (*chess.Move, float64) {
	var position = game.Position()
	var moves []*chess.Move = position.ValidMoves()

	if position.Status() == chess.Checkmate {
		return nil, math.Inf(-1)
	}

	if len(moves) == 0 {
		return nil, 0.0
	}

	var bestEvaluation float64 = math.Inf(-1)
	var bestMove *chess.Move

	orderMoves(moves, position)

	for _, move := range moves {
		var newPosition = position.Update(move)
		var evaluation float64 = -negamax(depth-1, math.Inf(-1), math.Inf(1), newPosition)

		if bestEvaluation <= evaluation {
			bestEvaluation = evaluation
			bestMove = move
		}
	}

	return bestMove, bestEvaluation
}

func playBestMove(depth uint8, game *chess.Game) (*chess.Move, float64) {
	var move, evaluation = findBestMove(depth, game)

	if move != nil {
		game.Move(move)
	}

	return move, evaluation
}
